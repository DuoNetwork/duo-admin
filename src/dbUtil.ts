import dynamoUtil from './database/dynamoUtil';
import sqlUtil from './database/sqlUtil';
import keyUtil from './keyUtil';
import { IOption, IPrice, ITrade } from './types';
import util from './util';
const localSQLauth = require('./keys/mysql.json');

class DbUtil {
	private dynamo: boolean = false;

	public async init(tool: string, option: IOption) {
		this.dynamo = option.dynamo;
		const role = util.getDynamoRole(tool, option.dynamo);
		const process = util.getStatusProcess(tool, option);
		util.log('role: ' + role);
		util.log('process: ' + process);

		dynamoUtil.init(option.live, role, process);
		if ((['bitfinex', 'gemini', 'kraken', 'gdax', 'commit'].includes(tool) && !option.dynamo) || option.server) {
			const sqlAuth = await keyUtil.getSqlAuth(option);
			sqlUtil.init(sqlAuth.host, sqlAuth.user, sqlAuth.password);
		} else
			sqlUtil.init(localSQLauth.host, localSQLauth.user, localSQLauth.password);
	}

	public insertTradeData(trade: ITrade, insertStatus: boolean) {
		return this.dynamo
			? dynamoUtil.insertTradeData(trade, insertStatus)
			: sqlUtil.insertTradeData(trade, insertStatus);
	}

	public insertPrice(price: IPrice) {
		return this.dynamo ? Promise.reject('invalid') : sqlUtil.insertPrice(price);
	}

	public readLastPrice(): Promise<IPrice> {
		return this.dynamo ? Promise.reject('invalid') : sqlUtil.readLastPrice();
	}

	public readSourceData(currentTimestamp: number): Promise<ITrade[]> {
		return this.dynamo ? Promise.reject('invalid') : sqlUtil.readSourceData(currentTimestamp);
	}

	public cleanDB(): Promise<void> {
		return this.dynamo ? Promise.reject() : sqlUtil.cleanDB();
	}

	public insertHeartbeat(data: object = {}): Promise<void> {
		return dynamoUtil.insertHeartbeat(data);
	}
}

const dbUtil = new DbUtil();
export default dbUtil;
