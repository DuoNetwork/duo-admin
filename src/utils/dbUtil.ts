import Web3Wrapper from '../../../duo-contract-wrapper/src/Web3Wrapper';
import * as CST from '../common/constants';
import { IOption, IPriceFix, ITrade } from '../common/types';
import dynamoUtil from './dynamoUtil';
import keyUtil from './keyUtil';
import sqlUtil from './sqlUtil';
import util from './util';

class DbUtil {
	public dynamo: boolean = false;

	public async init(tool: string, option: IOption, web3Wrapper: Web3Wrapper) {
		this.dynamo = option.dynamo;
		const process = util.getStatusProcess(tool, option);
		util.logInfo('process: ' + process);

		let config = {};
		try {
			config = require('../keys/aws/' + (option.live ? 'live' : 'dev') + '/admin.json');
		} catch (error) {
			util.logError(error);
		}
		await dynamoUtil.init(
			config,
			option.live,
			process,
			(value: string | number) => web3Wrapper.fromWei(value),
			async txHash => {
				const txReceipt = await web3Wrapper.getTransactionReceipt(txHash);
				if (!txReceipt) return null;
				return {
					status: txReceipt.status
				};
			}
		);
		if ([CST.TRADES, CST.COMMIT, CST.CLEAN_DB].includes(tool) && !option.dynamo) {
			const sqlAuth = await keyUtil.getSqlAuth(option);
			return sqlUtil.init(sqlAuth.host, sqlAuth.user, sqlAuth.password);
		}
	}

	public insertTradeData(trade: ITrade, insertStatus: boolean) {
		return this.dynamo
			? dynamoUtil.insertTradeData(trade, insertStatus)
			: sqlUtil.insertTradeData(trade, insertStatus);
	}

	public insertPrice(price: IPriceFix) {
		return this.dynamo ? Promise.reject('invalid') : sqlUtil.insertPrice(price);
	}

	public readLastPrice(quote: string, base: string): Promise<IPriceFix> {
		return this.dynamo ? Promise.reject('invalid') : sqlUtil.readLastPrice(quote, base);
	}

	public readSourceData(
		currentTimestamp: number,
		quote: string,
		base: string
	): Promise<ITrade[]> {
		return this.dynamo
			? Promise.reject('invalid')
			: sqlUtil.readSourceData(currentTimestamp, base, quote);
	}

	public cleanDB(): Promise<void> {
		return this.dynamo ? Promise.reject() : sqlUtil.cleanDB();
	}

	public insertHeartbeat(data: object = {}): Promise<void> {
		return dynamoUtil.insertHeartbeat(data);
	}
}

const dbUtil = new DbUtil();
export default dbUtil;
