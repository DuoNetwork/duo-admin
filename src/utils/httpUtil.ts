import request from 'request';

class HttpUtil {
	public get(url: string, headerOthers?: object): Promise<any> {
		return new Promise((resolve, reject) =>
			request(
				{
					url,
					headers: Object.assign(
						{
							'user-agent': 'node.js'
						},
						headerOthers || {}
					)
				},
				(error, res, body) => {
					if (error) reject(error);
					else if (res.statusCode === 200) resolve(body);
					else reject('Error status ' + res.statusCode + ' ' + res.statusMessage);
				}
			)
		);
	}

	public postJson(url: string, json: object): Promise<object> {
		return new Promise((resolve, reject) =>
			request(
				{
					url,
					method: 'POST',
					headers: {
						'Content-Type': 'application/json'
					},
					json
				},
				(error, res, body) => {
					if (error) reject(error);
					else if (res.statusCode === 200) resolve(body);
					else reject('Error status ' + res.statusCode + ' ' + res.statusMessage);
				}
			)
		);
	}
}

const httpUtil = new HttpUtil();
export default httpUtil;
