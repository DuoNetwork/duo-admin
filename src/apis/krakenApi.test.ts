// fix for @ledgerhq/hw-transport-u2f 4.28.0
import '@babel/polyfill';
import tradesRest from '../samples/kraken/tradesRest.json';
import dbUtil from '../utils/dbUtil';
import httpUtil from '../utils/httpUtil';
import api from './krakenApi';

api.init();
let sourceCashPair = ['ETH', 'USD']
	.sort(() => (api.settings.quoteInversed ? 1 : -1))
	.join(api.settings.separator);
if (api.settings.isLowercase) sourceCashPair = sourceCashPair.toLowerCase();
const localCashPair = 'ETH|USD';
api.sourcePairMapping[sourceCashPair] = localCashPair;

test('fetchTradesREST spot', async () => {
	api.settings.supportWS = false;
	api.last = {};
	api.tradeStatusLastUpdatedAt = {};

	httpUtil.get = jest.fn(() => Promise.resolve(JSON.stringify(tradesRest)));
	dbUtil.insertTradeData = jest.fn(() => Promise.resolve());

	await api.fetchTradesREST(sourceCashPair);
	expect((httpUtil.get as jest.Mock<Promise<void>>).mock.calls).toMatchSnapshot();
	const calls = (dbUtil.insertTradeData as jest.Mock<Promise<void>>).mock.calls;
	expect(calls).toMatchSnapshot();
	expect(calls.length).toEqual(
		Object.keys(tradesRest.result)
			.map(key => (tradesRest as { [key: string]: any }).result[key])
			.filter((x: any) => Array.isArray(x))
			.map(x => x.length)
			.reduce((x1: number, x2: number) => x1 + x2)
	);
});

test('fetchTradesWS', () => expect(() => api.fetchTradesWS()).toThrowErrorMatchingSnapshot());
