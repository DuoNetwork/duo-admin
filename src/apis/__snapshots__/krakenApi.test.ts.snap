// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchTradesREST spot 1`] = `
Array [
  Array [
    "https://api.kraken.com/0/public/Trades?pair=ETHUSD",
  ],
]
`;

exports[`fetchTradesREST spot 2`] = `
Array [
  Array [
    Object {
      "amount": 0.0839,
      "base": "USD",
      "id": "15335151284756",
      "price": 219.8835,
      "quote": "ETH",
      "source": "kraken",
      "timestamp": 1533515128475,
    },
    true,
  ],
  Array [
    Object {
      "amount": 1.31402601,
      "base": "USD",
      "id": "15335147744719",
      "price": 219.12,
      "quote": "ETH",
      "source": "kraken",
      "timestamp": 1533514774472,
    },
    false,
  ],
  Array [
    Object {
      "amount": 0.03243559,
      "base": "USD",
      "id": "15335147744656",
      "price": 219.66,
      "quote": "ETH",
      "source": "kraken",
      "timestamp": 1533514774465,
    },
    false,
  ],
  Array [
    Object {
      "amount": 0.0815328,
      "base": "USD",
      "id": "15335146199905",
      "price": 219.88,
      "quote": "ETH",
      "source": "kraken",
      "timestamp": 1533514619990,
    },
    false,
  ],
  Array [
    Object {
      "amount": 0.33962667,
      "base": "USD",
      "id": "15335145925313",
      "price": 219.88,
      "quote": "ETH",
      "source": "kraken",
      "timestamp": 1533514592531,
    },
    false,
  ],
  Array [
    Object {
      "amount": 0.00107193,
      "base": "USD",
      "id": "15335144787304",
      "price": 219.5,
      "quote": "ETH",
      "source": "kraken",
      "timestamp": 1533514478730,
    },
    false,
  ],
  Array [
    Object {
      "amount": 0.74665183,
      "base": "USD",
      "id": "15335144787284",
      "price": 219.99,
      "quote": "ETH",
      "source": "kraken",
      "timestamp": 1533514478728,
    },
    false,
  ],
  Array [
    Object {
      "amount": 0.21727624,
      "base": "USD",
      "id": "15335144787232",
      "price": 219.88,
      "quote": "ETH",
      "source": "kraken",
      "timestamp": 1533514478723,
    },
    false,
  ],
]
`;

exports[`fetchTradesWS 1`] = `"no ws"`;
